design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,791 @@
+```
+
+By registering a handler we have changed the corresponding `OpX` generic parameter.
+Fixing this error requires some non-trivial type gymnastic - I gave up after trying for ~15 minutes.
+
+### Cut them down: going from 2N+1 to 2 generic parameters
+
+The previous two examples should have convinced you that the 2N+1 generic parameters on `PokemonServiceBuilder` harm the ergonomics of our API.
+Can we get rid of them?
+
+Yes! Let's look at one possible approach:
+
+```rust
+pub struct PokemonServiceBuilder<Body, Plugin> {
+    check_health: Option<Route<Body>>,
+    do_nothing: Option<Route<Body>>,
+    get_pokemon_species: Option<Route<Body>>,
+    get_server_statistics: Option<Route<Body>>,
+    capture_pokemon: Option<Route<Body>>,
+    get_storage: Option<Route<Body>>,
+    plugin: Plugin,
+}
+```
+
+We no longer store the raw handlers inside `PokemonServiceBuilder`.
+We eagerly upgrade the operation handlers to a `Route` instance when they are registered with the builder.
+
+```rust
+impl<Body, Plugin> PokemonServiceBuilder<Body, Plugin> {
+    pub fn get_pokemon_species<Handler, Extensions>(mut self, handler: Handler) -> Self
+    /* Complex trait bounds */
+    {
+        let route = Route::new(Operation::from_handler(handler).upgrade(&self.plugin));
+        self.get_pokemon_species = Some(route);
+        self
+    }
+
+    /* other setters and methods */
+}
+```
+
+The existing API performs the upgrade when `build` is called, forcing `PokemonServiceBuilder` to store the raw handlers and keep two generic parameters around (`OpX` and `ExtsX`) for each operation.
+The proposed API requires plugins to be specified upfront, when creating an instance of the builder. They cannot be modified after a `PokemonServiceBuilder` instance has been built:
+
+```rust
+impl PokemonService<()> {
+    /// Constructs a builder for [`PokemonService`].
+    pub fn builder<Body, Plugin>(plugin: Plugin) -> PokemonServiceBuilder<Body, Plugin> {
+        PokemonServiceBuilder {
+            check_health: None,
+            do_nothing: None,
+            get_pokemon_species: None,
+            get_server_statistics: None,
+            capture_pokemon: None,
+            get_storage: None,
+            plugin,
+        }
+    }
+}
+```
+
+This constraint guarantees that all operation handlers are upgraded to a `Route` using the same set of plugins.
+
+Having to specify all plugins upfront is unlikely to have a negative impact on developers currently using `smithy-rs`.
+We have seen how cumbersome it is to break the startup logic into different functions using the current service builder API. Developers are most likely specifying all plugins and routes in the same function even if the current API allows them to intersperse route registrations and plugin registrations: they would simply have to re-order their registration statements to adopt the API proposed in this RFC.
+
+### Alternatives: allow new plugins to be registered after builder creation
```

Comment:
I think we should add a final alternative here:

- Don't solve the function signature problem at all, the customer just has to `PokemonService::builder()....build()` within one function.
- Solve the conditional branch problem with `Either<A, B>` types.

Reply:
> Don't solve the function signature problem at all, the customer just has to `PokemonService::builder()....build()` within one function.

I think the RFC should mention using macros instead of functions to workaround this limitation.

Reply:
Macros that we provide or macros written on an ad-hoc basis by the customer?

Reply:
I'd argue that writing a macro is a skill outside the comfort zone of most Rust users, who primarily consume them. Sketching a very simple `macro_rules!` is somewhat more accessible, but not much.